src/AsyncFunction.cs

```diff
@@ -0,0 +1,191 @@
+            }
+        }
+
+        /// <summary>
+        /// Check if this function has the given type signature
+        /// </summary>
+        /// <param name="returnType">Return type (use a tuple for multiple return types)</param>
+        /// <param name="parameters">The parameters of the function</param>
+        /// <returns>Returns true if the type signature of the function is valid or false if not.</returns>
+        public bool CheckTypeSignature(Type? returnType = null, params Type[] parameters)
+        {
+            return Function.CheckTypeSignature(returnType, parameters, Results, Parameters);
+        }
+
+        private async Task InvokeWithoutReturn(Value[] argsAndResults)
+        {
+            var trap = await Invoke(argsAndResults);
+
+            if (trap != IntPtr.Zero)
+            {
+                throw TrapException.FromOwnedTrap(trap);
+            }
+        }
+
+        private async Task<TResult?> InvokeWithReturn<TResult>(Value[] argsAndResults, IReturnTypeFactory<TResult> factory)
+        {
+            if (Results.Count == 0)
+            {
+                throw new InvalidOperationException($"Cannot call `InvokeWithReturn` on a method with 0 results");
+            }
+
+            var trap = await Invoke(argsAndResults.AsMemory());
+
+            return factory.Create(store.Context, store, trap, argsAndResults.AsSpan(Parameters.Count));
+        }
+
+        private async Task<IntPtr> Invoke(Memory<Value> argsAndResults)
+        {
+            using var pin = argsAndResults.Pin();
+
+            IntPtr trap;
+            IntPtr error;
+            nint futurePtr;
+            unsafe
+            {
+                var argsPtr = (Value*)pin.Pointer;
+                var resultsPtr = argsPtr + Parameters.Count;
+
+                if (Parameters.Count == 0)
+                {
+                    argsPtr = (Value*)0;
+                }
+
+                if (Results.Count == 0)
+                {
+                    resultsPtr = (Value*)0;
+                }
+
+                futurePtr = Native.wasmtime_func_call_async(store.Context.handle, func, argsPtr, (nuint)Parameters.Count, resultsPtr, (nuint)Results.Count, out trap, out error);
+
+                if (error != IntPtr.Zero)
+                {
+                    throw WasmtimeException.FromOwnedError(error);
+                }
+
+                if (trap != IntPtr.Zero)
+                {
+                    return trap;
+                }
+            }
+
+            try
+            {
+                while (!CallFuture.Native.wasmtime_call_future_poll(futurePtr))
+                {
+                    if (error != IntPtr.Zero)
+                    {
+                        throw WasmtimeException.FromOwnedError(error);
+                    }
+
+                    if (trap != IntPtr.Zero)
+                    {
+                        return trap;
+                    }
+
+                    await Task.Yield();
```

Comment:
It looks like this will produce high CPU use by calling `Task.Yield()` in a loop in order to check when the future has completed. I haven't looked much into how the new async support works in the Wasmtime C API, but maybe is there a different way of get notified when a future might have completed?

From a quick look at the `async.h` documentation, I was thinking that a returned `wasmtime_call_future_t` might completed once one of the async callback functions has completed. Maybe it would be possible to trigger checking of the `wasmtime_call_future_t` once one of the `wasmtime_async_continuation_t` is set to completed (as I guess the call of `wasmtime_call_future_poll` will trigger Wasmtime to call the open `wasmtime_func_async_continuation_callback_t` callbacks.